xds/internal/testutils/fakeclient/client.go

```diff
@@ -251,14 +268,15 @@ func NewClientWithName(name string) *Client {
 		name:         name,
 		ldsWatchCh:   testutils.NewChannel(),
 		rdsWatchCh:   testutils.NewChannel(),
-		cdsWatchCh:   testutils.NewChannel(),
+		cdsWatchCh:   testutils.NewChannelWithSize(10),
```

Comment:
This is OK in general.
My concern is that we may miss some failures that could be caught by this (though it's not by design :man_shrugging:)

The size one channel has the nice ability that, an unexpected write would block, and cause the test to timeout.

Now, for example, if there's a bug, an aggregated cluster `{a,b}` results in 3 watches for `a, b, a`, we _may_ miss it. 
An easy way is, after `WaitForWatch(a), WaitForWatch(b)`, do another `WaitForWatch()`, and expect an timeout error.
The same goes for `WaitForCancel()`.

I will probably pick one test (`TestSwitchClusterNodeBetweenLeafAndAggregated`) to add these checks, as it covers both `WaitForWatch()` and `WaitForCancel()`.

Reply:
I'll add these validations, but that seems like overkill. Just like my previous comment, the main validation I feel like matters is the update passed to cds. If the handler started extra watches it wouldn't match/the update wouldn't actually build (as callbacks haven't been invoked yet because I do that manually).

Reply:
Another thing I could do is scale the buffer down, but that would a. couple the check (logical check of timing out not explicit check) with the number of children and also there would still be a problem on cases where you have children size < buffer size. I'm starting to really agree with your thought process here.